agent/rpc/peering/service.go

```diff
@@ -463,15 +463,23 @@ func (s *Server) Establish(
 	return resp, nil
 }
 
-// validatePeeringInPartition makes sure that we don't create a peering in the same partition. We validate by looking at
-// the remotePeerID from the PeeringToken and looking up for a peering in the partition. If there is one and the
-// request partition is the same, then we are attempting to peer within the partition, which we shouldn't.
-func (s *Server) validatePeeringInPartition(remotePeerID, partition string) error {
-	_, peering, err := s.Backend.Store().PeeringReadByID(nil, remotePeerID)
+// validatePeeringLocality makes sure that we don't create a peering in the cluster/partition it was generated.
+// We validate by looking at the remotePeerID from the PeeringToken and looking up for a peering in the partition.
+// If there is one and the request partition is the same, then we are attempting to peer within the partition, which we shouldn't.
+// We also perform a check to verify if the ServerName of the PeeringToken overlaps with our own, we do not process it
+// unless we've been able to find the peering in the store, i.e. this peering is between two local partitions.
+func (s *Server) validatePeeringLocality(token *structs.PeeringToken, partition string) error {
+	_, peering, err := s.Backend.Store().PeeringReadByID(nil, token.PeerID)
 	if err != nil {
 		return fmt.Errorf("cannot read peering by ID: %w", err)
 	}
 
+	// If the token has the same server name as thing cluster, but we can't find the peering
+	// in our store, it indicates a naming conflict.
+	if s.Backend.GetServerName() == token.ServerName && peering == nil {
+		return fmt.Errorf("conflict - peering token \"ServerName\" matches the current cluster but was not generated here")
+	}
```

Comment:
We could also hit this scenario if an operator deletes a peering after generating a token, but later tries to establish within the same cluster with that token.

Reply:
We could also expose the server name with `%q` for easier debugging in the log message

Reply:
Hmm, I see your point. I don't really like the either-or message because we should be able to point operators to the real problem, but after staring at this for a while I don't know a better way. I also don't want to optimize too much for the same-cluster case (which seems like the degenerate case). How about tweaking it just a bit:
`conflict - peering token \"ServerName\" matches the current cluster \"ServerName\", %q, but there is no record in the database`

Reply:
Yeah it sounds more general 👍 

Styling-wise I prefer keeping server name as a normal noun (e.g. peering token's server name does not match the current cluster's server name) but not a blocker

Reply:
Normal-case works for me. 🚀